src/app/CommandHandler.h

```diff
@@ -327,12 +351,29 @@ class CommandHandler
      * @param [in] aData the data for the response.
      */
     template <typename CommandData>
-    CHIP_ERROR AddResponseData(const ConcreteCommandPath & aRequestCommandPath, const CommandData & aData)
+    inline CHIP_ERROR AddResponseData(const ConcreteCommandPath & aRequestCommandPath, const CommandData & aData)
+    {
+        DataModelEncoderToTLV<CommandData> encoder(aData);
+        return AddResponseDataViaEncoder(aRequestCommandPath, CommandData::GetCommandId(), encoder);
+    }
+
+    /**
+     * API for adding a data response.  The encoded is generally expected to encode
+     * a ClusterName::Commands::CommandName::Type struct, but any
+     * object should work.
+     *
+     * @param [in] aRequestCommandPath the concrete path of the command we are
+     *             responding to.
+     * @param [in] commandId the command whose content is being encoded.
+     * @param [in] aData the data for the response.
+     */
+    CHIP_ERROR AddResponseDataViaEncoder(const ConcreteCommandPath & aRequestCommandPath, CommandId commandId,
```

Comment:
This is a new public function - is the intent that devs start moving away from AddResponseData? ie, if you're a dev, which one of these should you choose?

Reply:
Added comments. I think `AddResponse/AddResponseData` is what people will use because of the auto `GetCommandId()` calls.

I kept this public because I expect I will split this to become virtual while the templated bits will stay templated (and non-virtual).